File: CSharp/CsLibrary/String.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace JavaClasses
{

    /// <summary>
    /// Class Object is the root of the class hierarchy.
    /// </summary>
    public class Object
    {

        /// <summary>
        /// Indicates whether some other object is "equal to" this one.
        /// </summary>
        /// <param name="obj">the reference object with which to compare.</param>
        /// <returns>true if this object is the same as the obj argument; false otherwise.</returns>
        public virtual bool equals(Object obj)
        {
            return this == obj;
        }

        /// <summary>
        /// Returns a string representation of the object.
        /// </summary>
        /// <returns>a string representation of the object.</returns>
        public virtual String toString()
        {
            return new String(base.ToString());
        }

        public override bool Equals(object obj)
        {
            return equals(obj as Object);
        }

        public override string ToString()
        {
            return toString().s;
        }

    }

    /// <summary>
    /// This interface imposes a total ordering on the objects of each class that implements it.
    /// </summary>
    /// <typeparam name="T"></typeparam>
    public abstract class Comparable<T> : Object, IComparable<T> where T : Object
    {

        /// <summary>
        /// Compares this object with the specified object for order.
        /// </summary>
        /// <param name="o">the object to be compared.</param>
        /// <returns>a negative integer, zero, or a positive integer as this object
        /// is less than, equal to, or greater than the specified object.</returns>
        public abstract int compareTo(T o);

        public int CompareTo(T o)
        {
            return compareTo(o);
        }

    }

    /// <summary>
    /// A comparison function, which imposes a total ordering on some collection of objects.
    /// </summary>
    /// <typeparam name="T"></typeparam>
    public abstract class Comparator<T> : Object, IComparer<T> where T : Object
    {

        /// <summary>
        /// Compares its two arguments for order.
        /// </summary>
        /// <param name="o1">the first object to be compared.</param>
        /// <param name="o2">the second object to be compared.</param>
        /// <returns>negative integer, zero, or a positive integer as the
        /// first argument is less than, equal to, or greater than the second.</returns>
        public abstract int compare(T o1, T o2);

        public int Compare(T o1, T o2)
        {
            return compare(o1, o2);
        }

    }

    /// <summary>
    /// The String class represents character strings.
    /// </summary>
    /// <author>Petr Kozler (A13B0359P)</author>
    public class String : Comparable<String>
    {

        public string s { get; private set; }

        /// <summary>
        /// Constructs a new String by decoding the specified array of bytes using
        /// the platform's default charset.
        /// </summary>
        /// <param name="bytes">The bytes to be decoded into characters
        /// </param>
        public String(sbyte[] bytes)
        {
            // parsing to the unsigned byte array
            byte[] unsigned = (byte[])(Array)bytes;
            s = Encoding.Default.GetString(unsigned);
        }

        /// <summary>
        /// Constructs a new String by decoding the specified subarray of bytes using
        /// the platform's default charset.
        /// </summary>
        /// <param name="bytes">The bytes to be decoded into characters
        /// </param><param name="offset">The index of the first byte to decode
        /// </param><param name="length">The number of bytes to decode
        /// </param>
        public String(sbyte[] bytes, int offset, int length)
        {
            byte[] unsigned = (byte[])(Array)bytes;
            s = Encoding.Default.GetString(unsigned, offset, length);
        }

        /// <summary>
        /// Initializes a newly created String object so that it represents the same
        /// sequence of characters as the argument; in other words, the newly created
        /// string is a copy of the argument string. Unless an explicit copy of
        /// original is needed, use of this constructor is unnecessary since Strings
        /// are immutable.
        /// </summary>
        /// <param name="original">A String
        /// </param>
        public String(string original)
        {
            s = original;
        }

        /// <summary>
        /// Compares two strings lexicographically.
        /// </summary>
        /// <param name="anotherString">the String to be compared.
        /// </param><returns>the value 0 if the argument string is equal to this string; a
        /// value less than 0 if this string is lexicographically less than the
        /// string argument; and a value greater than 0 if this string is
        /// lexicographically greater than the string argument.
        /// </returns>
        public override int compareTo(String anotherString)
        {
            return s.CompareTo(anotherString.s);
        }

        /// <summary>
        /// Compares this string to the specified object.
        /// </summary>
        /// <param name="anObject">The object to compare this String against
        /// </param><returns>true if the given object represents a String equivalent to this
        /// string, false otherwise
        /// </returns>
        public override bool equals(Object anObject)
        {
            if (anObject == null)
            {
                return false;
            }

            if (GetType() != anObject.GetType())
            {
                return false;
            }

            return s.Equals((anObject as String).s);
        }

        /// <summary>
        /// Returns a new string that is a substring of this string.
        /// </summary>
        /// <param name="beginIndex">the beginning index, inclusive.
        /// </param><returns>the specified substring.
        /// </returns>
        public String substring(int beginIndex)
        {
            return new String(s.Substring(beginIndex));
        }

        /// <summary>
        /// Returns a new string that is a substring of this string.
        /// </summary>
        /// <param name="beginIndex">the beginning index, inclusive.
        /// </param><param name="endIndex">the ending index, exclusive.
        /// </param><returns>the specified substring.
        /// </returns>
        public String substring(int beginIndex, int endIndex)
        {
            // calculating the length from the specified end index
            return new String(s.Substring(beginIndex, endIndex - beginIndex));
        }

        /// <summary>
        /// Returns the index within this string of the first occurrence of the
        /// specified character.
        /// </summary>
        /// <param name="ch">a character (Unicode code point).
        /// </param><returns>the index of the first occurrence of the character in the
        /// character sequence represented by this object, or -1 if the character
        /// does not occur.
        /// </returns>
        public int indexOf(int ch)
        {
            return s.IndexOf((char)ch);
        }

        /// <summary>
        /// Returns the index within this string of the first occurrence of the
        /// specified character, starting the search at the specified index.
        /// </summary>
        /// <param name="ch">a character (Unicode code point).
        /// </param><param name="fromIndex">the index to start the search from.
        /// </param><returns>the index of the first occurrence of the character in the
        /// character sequence represented by this object that is greater than or
        /// equal to fromIndex, or -1 if the character does not occur.
        /// </returns>
        public int indexOf(int ch, int fromIndex)
        {
            return s.IndexOf((char)ch, fromIndex);
        }

        /// <summary>
        /// Returns the index within this string of the first occurrence of the
        /// specified substring.
        /// </summary>
        /// <param name="str">the substring to search for.
        /// </param><returns>the index of the first occurrence of the specified substring, or
        /// -1 if there is no such occurrence.
        /// </returns>
        public int indexOf(String str)
        {
            return s.IndexOf(str.s);
        }

        /// <summary>
        /// Returns the index within this string of the first occurrence of the
        /// specified substring, starting at the specified index.
        /// </summary>
        /// <param name="str">the substring to search for.
        /// </param><param name="fromIndex">the index from which to start the search.
        /// </param><returns>the index of the first occurrence of the specified substring,
        /// starting at the specified index, or -1 if there is no such occurrence.
        /// </returns>
        public int indexOf(String str, int fromIndex)
        {
            return s.IndexOf(str.s, fromIndex);
        }

        /// <summary>
        /// Returns the length of this string.
        /// </summary>
        /// <returns>the length of the sequence of characters represented by this
        /// object.
        /// </returns>
        public int length()
        {
            return s.Length;
        }

        /// <summary>
        /// Returns a copy of the string, with leading and trailing whitespace
        /// omitted.
        /// </summary>
        /// <returns>A copy of this string with leading and trailing white space
        /// removed, or this string if it has no leading or trailing white space.
        /// </returns>
        public String trim()
        {
            return new String(s.Trim());
        }

        /// <summary>
        /// Converts all of the characters in this String to lower case using the
        /// rules of the default locale.
        /// </summary>
        /// <returns>the String, converted to lowercase.
        /// </returns>
        public String toLowerCase()
        {
            return new String(s.ToLower());
        }

        /// <summary>
        /// Converts all of the characters in this String to upper case using the
        /// rules of the default locale.
        /// </summary>
        /// <returns>the String, converted to uppercase.
        /// </returns>
        public String toUpperCase()
        {
            return new String(s.ToUpper());
        }

        /// <summary>
        /// Returns the char value at the specified index.
        /// </summary>
        /// <param name="index">the index of the char value.
        /// </param><returns>the char value at the specified index of this string. The first
        /// char value is at index 0.
        /// </returns>
        public char charAt(int index)
        {
            return s[index];
        }

        /// <summary>
        /// Returns a new string resulting from replacing all occurrences of oldChar
        /// in this string with newChar.
        /// </summary>
        /// <param name="oldChar">the old character.
        /// </param><param name="newChar">the new character.
        /// </param><returns>a string derived from this string by replacing every occurrence
        /// of oldChar with newChar.
        /// </returns>
        public String replace(char oldChar, char newChar)
        {
            return new String(s.Replace(oldChar, newChar));
        }

        /// <summary>
        /// Tests if this string starts with the specified prefix.
        /// </summary>
        /// <param name="prefix">the prefix.
        /// </param><returns>true if the character sequence represented by the argument is a
        /// prefix of the character sequence represented by this string; false
        /// otherwise. Note also that true will be returned if the argument is an
        /// empty string or is equal to this String object as determined by the
        /// equals(Object) method.
        /// </returns>
        public bool startsWith(String prefix)
        {
            return s.StartsWith(prefix.s);
        }

        /// <summary>
        /// Tests if this string ends with the specified suffix.
        /// </summary>
        /// <param name="suffix">the suffix.
        /// </param><returns>true if the character sequence represented by the argument is a
        /// suffix of the character sequence represented by this object; false
        /// otherwise. Note that the result will be true if the argument is the empty
        /// string or is equal to this String object as determined by the
        /// equals(Object) method.
        /// </returns>
        public bool endsWith(String suffix)
        {
            return s.EndsWith(suffix.s);
        }

        /// <summary>
        /// Returns true if, and only if, length() is 0.
        /// </summary>
        /// <returns>true if length() is 0, otherwise false
        /// </returns>
        public bool isEmpty()
        {
            return s.Length == 0;
        }

        /// <summary>
        /// This object (which is already a string!) is itself returned.
        /// </summary>
        /// <returns>the string itself.</returns>
        public override String toString()
        {
            return new String(s);
        }
        
        public static implicit operator String(string original)
        {
            return new String(original);
        }
        
    }

}

File: CPlusPlus/CppLibrary/String.cpp
#include "String.h"

#include <algorithm>
#include <string>
#include <iostream>
#include <sstream>
#include <iomanip>

/**
 * Indicates whether some other object is "equal to" this one.
 *
 * @param obj the reference object with which to compare.
 * @return true if this object is the same as the obj argument; false otherwise.
 */
bool Object::equals(Object *obj) {
    return this == obj;
}

/**
 * Returns a string representation of the object.
 *
 * @return a string representation of the object.
 */
String *Object::toString() {
    std::stringstream stream;
    stream << std::hex << this;
    return new String(stream.str());
}

std::ostream &operator<<(std::ostream &s, Object &o) {
    return s << o.toString()->_s();
}

/**
 * Constructs a new String by decoding the specified array of bytes using
 * the platform's default charset.
 *
 * @param bytes The bytes to be decoded into characters
 */
String::String(int8_t *value, int32_t length) {
    this->s = ((char *) value, length);
}

/**
 * Constructs a new String by decoding the specified subarray of bytes using
 * the platform's default charset.
 *
 * @param bytes The bytes to be decoded into characters
 * @param offset The index of the first byte to decode
 * @param length The number of bytes to decode
 */
String::String(int8_t *value, int32_t offset, int32_t length) {
    this->s = ((char *) (value + offset), length);
}

/**
 * Initializes a newly created String object so that it represents the same
 * sequence of characters as the argument; in other words, the newly created
 * string is a copy of the argument string. Unless an explicit copy of
 * original is needed, use of this constructor is unnecessary since Strings
 * are immutable.
 *
 * @param original A String
 */
String::String(std::string original) {
    this->s = original;
}

/**
 * Compares two strings lexicographically.
 *
 * @param anotherString the String to be compared.
 * @return the value 0 if the argument string is equal to this string; a
 * value less than 0 if this string is lexicographically less than the
 * string argument; and a value greater than 0 if this string is
 * lexicographically greater than the string argument.
 */
int32_t String::compareTo(String *anotherString) {
    return this->s.compare(anotherString->s);
}

/**
 * Compares this string to the specified object.
 *
 * @param anObject The object to compare this String against
 * @return true if the given object represents a String equivalent to this
 * string, false otherwise
 */
bool String::equals(String *anObject) {
    if (anObject == nullptr) {
        return false;
    }

    if (sizeof (*this) != sizeof (*anObject)) {
        return false;
    }

    return (bool) !this->s.compare(anObject->s);
}

/**
 * Returns a new string that is a substring of this string.
 * 
 * @param beginIndex the beginning index, inclusive.
 * @return the specified substring.
 */
String *String::substring(int32_t beginIndex) {
    return new String(this->s.substr(beginIndex));
}

/**
 * Returns a new string that is a substring of this string.
 *
 * @param beginIndex the beginning index, inclusive.
 * @param endIndex the ending index, exclusive.
 * @return the specified substring.
 */
String *String::substring(int32_t beginIndex, int32_t endIndex) {
    return new String(this->s.substr(beginIndex, endIndex - beginIndex + 1));
}

/**
 * Returns the index within this string of the first occurrence of the
 * specified character.
 *
 * @param ch a character (Unicode code point).
 * @return the index of the first occurrence of the character in the
 * character sequence represented by this object, or -1 if the character
 * does not occur.
 */
int32_t String::indexOf(int32_t ch) {
    return this->s.find((char) ch);
}

/**
 * Returns the index within this string of the first occurrence of the
 * specified character, starting the search at the specified index.
 *
 * @param ch a character (Unicode code point).
 * @param fromIndex the index to start the search from.
 * @return the index of the first occurrence of the character in the
 * character sequence represented by this object that is greater than or
 * equal to fromIndex, or -1 if the character does not occur.
 */
int32_t String::indexOf(int32_t ch, int32_t fromIndex) {
    return this->s.find((char) ch, fromIndex);
}

/**
 * Returns the index within this string of the first occurrence of the
 * specified substring.
 *
 * @param str the substring to search for.
 * @return the index of the first occurrence of the specified substring, or
 * -1 if there is no such occurrence.
 */
int32_t String::indexOf(String *str) {
    return this->s.find(str->s);
}

/**
 * Returns the index within this string of the first occurrence of the
 * specified substring, starting at the specified index.
 *
 * @param str the substring to search for.
 * @param fromIndex the index from which to start the search.
 * @return the index of the first occurrence of the specified substring,
 * starting at the specified index, or -1 if there is no such occurrence.
 */
int32_t String::indexOf(String *str, int32_t fromIndex) {
    return this->s.find(str->s, fromIndex);
}

/**
 * Returns the length of this string.
 *
 * @return the length of the sequence of characters represented by this
 * object.
 */
int32_t String::length() {
    return this->s.length();
}

/**
 * Returns a copy of the string, with leading and trailing whitespace
 * omitted.
 *
 * @return A copy of this string with leading and trailing white space
 * removed, or this string if it has no leading or trailing white space.
 */
String *String::trim() {
    const char* white_spaces = " \t\n\r\f";

    int32_t first = this->s.find_first_not_of(white_spaces);

    if (first == std::string::npos) {
        return new String(std::string(""));
    }

    int32_t last = this->s.find_last_not_of(white_spaces);

    return new String(this->s.substr(first, last - first + 1));
}

/**
 * Converts all of the characters in this String to lower case using the
 * rules of the default locale.
 *
 * @return the String, converted to lowercase.
 */
String *String::toLowerCase() {
    std::string lowerCaseStr = this->s;
    std::transform(this->s.begin(), this->s.end(), std::back_inserter(lowerCaseStr), ::tolower);
    return new String(lowerCaseStr);
}

/**
 * Converts all of the characters in this String to upper case using the
 * rules of the default locale.
 *
 * @return the String, converted to uppercase.
 */
String *String::toUpperCase() {
    std::string upperCaseStr = this->s;
    std::transform(this->s.begin(), this->s.end(), std::back_inserter(upperCaseStr), ::toupper);
    return new String(upperCaseStr);
}

/**
 * Returns the char value at the specified index.
 *
 * @param index the index of the char value.
 * @return the char value at the specified index of this string. The first
 * char value is at index 0.
 */
char String::charAt(int32_t index) {
    return this->s.at(index);
}

/**
 * Returns a new string resulting from replacing all occurrences of oldChar
 * in this string with newChar.
 *
 * @param oldChar the old character.
 * @param newChar the new character.
 * @return a string derived from this string by replacing every occurrence
 * of oldChar with newChar.
 */
String *String::replace(char oldChar, char newChar) {
    std::string str = this->s;
    std::replace(str.begin(), str.end(), oldChar, newChar);

    return new String(str);
}

/**
 * Tests if this string starts with the specified prefix.
 *
 * @param prefix the prefix.
 * @return true if the character sequence represented by the argument is a
 * prefix of the character sequence represented by this string; false
 * otherwise. Note also that true will be returned if the argument is an
 * empty string or is equal to this String object as determined by the
 * equals(Object) method.
 */
bool String::startsWith(String *prefix) {
    return (s.size() >= prefix->_s().size() && s.find(prefix->_s()) == 0);
}

/**
 * Tests if this string ends with the specified suffix.
 *
 * @param suffix the suffix.
 * @return true if the character sequence represented by the argument is a
 * suffix of the character sequence represented by this object; false
 * otherwise. Note that the result will be true if the argument is the empty
 * string or is equal to this String object as determined by the
 * equals(Object) method.
 */
bool String::endsWith(String *suffix) {
    return (s.size() >= suffix->_s().size() &&
            s.compare(s.size() - suffix->_s().size(), suffix->_s().size(), suffix->_s()) == 0);
}

/**
 * Returns true if, and only if, length() is 0.
 *
 * @return true if length() is 0, otherwise false
 */
bool String::isEmpty() {
    return (this->s.length() == 0);
}

/**
 * This object (which is already a string!) is itself returned.
 *
 * @return the string itself.
 */
String *String::toString() {
    return new String(s);
}

String *String::operator=(std::string &s) {
    return new String(s);
}
